File: lib/providers/hn-algolia.ts
/**
 * Hacker News search via the Algolia-hosted HN API.
 *
 * No auth, no rate limit headers worth worrying about. Useful for
 * finding founders / indie hackers / technical CEOs who post on HN.
 *
 * API docs: https://hn.algolia.com/api
 */

import type { ProspectCandidate } from "./brave-search"

const HN_SEARCH = "https://hn.algolia.com/api/v1/search"

interface HnHit {
  objectID: string
  author: string
  title: string | null
  story_title: string | null
  url: string | null
  story_url: string | null
  comment_text: string | null
  created_at: string
}

/**
 * Returns one prospect per *unique HN author* matching the query.
 * The author handle is the closest thing HN exposes to a person; we
 * use their HN profile URL as the source_url and the matched story/
 * comment snippet as their context.
 */
export async function searchHnUsers(
  query: string,
  maxResults: number,
): Promise<ProspectCandidate[]> {
  const url = `${HN_SEARCH}?query=${encodeURIComponent(query)}&hitsPerPage=${Math.min(maxResults * 3, 100)}&tags=story,comment`
  let res: Response
  try {
    res = await fetch(url, { signal: AbortSignal.timeout(10_000) })
  } catch {
    return []
  }
  if (!res.ok) return []
  const data = (await res.json()) as { hits?: HnHit[] }
  const hits = data.hits ?? []

  // Dedupe by author — one row per person, picking the best hit.
  const byAuthor = new Map<string, HnHit>()
  for (const h of hits) {
    if (!h.author) continue
    if (!byAuthor.has(h.author)) byAuthor.set(h.author, h)
  }

  const out: ProspectCandidate[] = []
  for (const h of Array.from(byAuthor.values()).slice(0, maxResults)) {
    const snippet =
      h.story_title ?? h.title ?? stripHtml(h.comment_text ?? "") ?? `HN activity for ${h.author}`
    out.push({
      name: h.author,
      title: "Active on Hacker News",
      company: "(independent)",
      source: "hn",
      source_url: `https://news.ycombinator.com/user?id=${encodeURIComponent(h.author)}`,
      snippet: truncate(snippet, 240),
    })
  }
  return out
}

function stripHtml(s: string): string {
  return s.replace(/<[^>]+>/g, "").replace(/&[#\w]+;/g, " ").trim()
}

function truncate(s: string, n: number): string {
  return s.length <= n ? s : s.slice(0, n - 1) + "…"
}

File: backend-python/leadgen_backend/providers/hn_algolia.py
import urllib.parse
import httpx
import re

HN_SEARCH = "https://hn.algolia.com/api/v1/search"

def strip_html(s: str) -> str:
    s = re.sub(r'<[^>]+>', '', s)
    s = re.sub(r'&[#\w]+;', ' ', s)
    return s.strip()

def truncate(s: str, n: int) -> str:
    return s if len(s) <= n else s[:n-1] + "…"

async def search_hn_users(query: str, max_results: int) -> list[dict]:
    params = {
        "query": query,
        "hitsPerPage": min(max_results * 3, 100),
        "tags": "story,comment"
    }
    
    try:
        async with httpx.AsyncClient(timeout=10) as client:
            res = await client.get(HN_SEARCH, params=params)
            if res.status_code >= 400: return []
            data = res.json()
    except httpx.HTTPError:
        return []
        
    hits = data.get("hits", [])
    
    by_author = {}
    for h in hits:
        author = h.get("author")
        if not author: continue
        if author not in by_author:
            by_author[author] = h
            
    out = []
    for h in list(by_author.values())[:max_results]:
        author = h["author"]
        snippet_raw = h.get("story_title") or h.get("title") or strip_html(h.get("comment_text") or "") or f"HN activity for {author}"
        
        out.append({
            "name": author,
            "title": "Active on Hacker News",
            "company": "(independent)",
            "source": "hn",
            "source_url": f"https://news.ycombinator.com/user?id={urllib.parse.quote(author)}",
            "snippet": truncate(snippet_raw, 240)
        })
        
    return out
